### ISSUE TITLE
Investigate performance of Antiforgery

### ISSUE THREAD

**Issue by BrennanConroy** [dotnet/aspnetcore#50065]
- [ ] We should add some antiforgery E2E benchmarks to our [regular aspnet/benchmark](https://github.com/aspnet/Benchmarks/tree/main/scenarios) runs
- [ ] Find and fix low-hanging fruit
  - [ ] e.g. https://github.com/dotnet/aspnetcore/blob/50c956218794bad8dce88609f4c2e4e711b739d1/src/Antiforgery/src/Internal/DefaultAntiforgeryTokenGenerator.cs#L62 and https://github.com/dotnet/aspnetcore/blob/50c956218794bad8dce88609f4c2e4e711b739d1/src/Antiforgery/src/Internal/DefaultAntiforgeryTokenGenerator.cs#L145
- [ ] Affect on Routing especially for non-form POSTs https://github.com/dotnet/aspnetcore/pull/50016#issuecomment-1677823427

**Comment by yunusefendi52** [comment]
I am seeing some performance bottleneck using when using antiforgery as well, I trace using dotnet trace.
<img width="700" alt="Screenshot 2023-08-20 at 1 48 13 PM" src="https://github.com/dotnet/aspnetcore/assets/19399214/3ae28abe-055f-4fd8-9afd-57d4b2454ac4">

Not sure if it's really because of antiforgery it self or data protection

**Comment by Westboldyi** [comment]
Startup of blazor server app with `UseAntiforgery()` causes clients to all timeout for the first 2 minutes or so. Threads just idle and blocked and client connections timeout. Removing `UseAntiforgery()` and disabling it alleviates this issue `app.MapRazorComponents<App>().DisableAntiforgery()`

![Image](https://github.com/user-attachments/assets/08888028-05d2-49dc-9254-a4d022d907a0)

**Comment by DeagleGross** [comment]
I am looking into Antiforgery improvements, and I can see many of allocations can be avoided, but only if DataProtection layer supports overloads with `Span<byte>`. I have done a review of the code and here are necessary APIs which have to be exposed and implemented to take advantage of them in Antiforgery:
https://github.com/dotnet/aspnetcore/issues/44758#issuecomment-2893933499

**Comment by DeagleGross** [comment]
> Startup of blazor server app with `UseAntiforgery()` causes clients to all timeout for the first 2 minutes or so. Threads just idle and blocked and client connections timeout. Removing `UseAntiforgery()` and disabling it alleviates this issue `app.MapRazorComponents<App>().DisableAntiforgery()`
> 
> ![Image](https://github.com/user-attachments/assets/08888028-05d2-49dc-9254-a4d022d907a0)

@Westboldyi could you please [submit a new issue](https://github.com/dotnet/aspnetcore/issues) and make a minimal repro? This seems like a bug and should be investigated separately from the performance

**Comment by captainsafia** [comment]
@DeagleGross @BrennanConroy Do we feel like we've done enough in .NET 10 to call this "done" or should be move this to the backlog for further work?

**Comment by DeagleGross** [comment]
@captainsafia the plan is to complete this in .NET 11 already. There are still improvements to be done in `DataProtection` and currently are in progress (#62903), and the follow-up in Antiforgery is required.

Honestly, not sure we should keed `area-mvc` on the issue (I think you came here because of area, right?). Antiforgery is used in many places - i.e. blazor. We should just keep `area-perf` here. Let me know if that makes sense

**Comment by captainsafia** [comment]
Yep, I came here because of the area-mvc label. Agree that marking it area-perf is probably better. I'll tentiatively put this in .NET 11 Planning as well.

**Comment by DeagleGross** [comment]
I guess we can close this:
1) made an improvement for existing Antiforgery - see perf results here https://github.com/dotnet/aspnetcore/issues/65361#issue-3913374837
2) if we decide to go with new CSRF algorithm, it will also be an improvement. See related issue https://github.com/dotnet/aspnetcore/issues/65127
